### PR TITLE
QA-2170: Lift and shift to use semver versioning for contract tests

### PR DESCRIPTION
**Pull request by ichengchang**
Jira ticket: [QA-2170](https://broadworkbench.atlassian.net/browse/QA-2170)

[QA-2170]: https://broadworkbench.atlassian.net/browse/QA-2170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

More details on why we're doing this shifting. Considering the creation of a ledger to trace contracts through PR and merge commits. It is also crucial for Sherlock to be able to interact with Pact Broker to record deployment environment for the appropriate release tag.

1.     Multiple Developers and Commits in a PR:
- Multiple developers can collaborate on the same pull request (PR).
- The PR can have multiple commits.

2.     Sherlock and Helm Charts for `datarepo`:
- Sherlock doesn't synchronize Helm charts for the `datarepo` on PR commits
- The synchronization is only called upon during merge commit to `develop` branch

3.     Continuous Publishing and Verification of Contracts:
- You still want contracts from the PR to be continually published or verified.
- The tagging mechanism is introduced to create a ledger of published and verified contracts in Pact Broker. For PRs, the tag is appended with the commit hash. When the live tag is pushed to GitHub during merge commit, the contracts will be published or verified using the live tag without hash.
- This ledger allows developers to trace issues on both the consumer or provider side.

### CODE REVIEW

File: .github/workflows/verify_consumer_pacts.yaml
Comment (okotsopoulos) [inline .github/workflows/verify_consumer_pacts.yaml]: Did you find that this addition was necessary?  We set the Pact broker URL in our application properties, so I think this is redundant.

File: .github/actions/bump-skip/action.yml
Comment (okotsopoulos) [inline .github/actions/bump-skip/action.yml:4]: I'm confused by this authorship, should it be removed or updated to you?

File: .github/workflows/verify_consumer_pacts.yaml
Comment (okotsopoulos) [inline .github/workflows/verify_consumer_pacts.yaml]: I want to note that TDR merges its feature branches into `develop` branch, not `main`.  Is this correct?

File: src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java
Comment (okotsopoulos) [inline src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java]: This could be a PR or a local run.

File: .github/workflows/tag.yml
Comment (okotsopoulos) [inline .github/workflows/tag.yml:69]: Want to make sure I understand:

In the last GHA run kicked off by commit `789dd32…`, this printed tag was `1.556.0-789dd32`.  The suffix is the commit name, which looks like it will help distinguish between multiple runs made when the live tag has not changed.

But is Sherlock okay with this?  I don't think our actual tag and deploy GHA suffixes its tags this way.

**Review by okotsopoulos — COMMENTED** [review]
Thank you so much for working on this.  To the best of my ability, I've reviewed and think it looks good.

It is definitely a pain point that datarepo maintains many of its own GHA helpers rather than ideally sharing those common among the rest of the org.  But remedying this is OOS for this work.

Could you please flesh out the PR description to explain why we're undertaking this change?  Notable is that datarepo's movement to use `semver` versioning for contract tests is primarily for consistency with how our app versions are reported to Sherlock, and not just because `semver` is a lot more user-friendly.

Could you please show how these changes resulted in the desired Pact broker interactions and meaningful `can-i-deploy` runs?  I ask because it was not abundantly clear to me before as a developer what was broken about the old set-up.

File: .github/workflows/verify_consumer_pacts.yaml
Comment (ichengchang) [reply .github/workflows/verify_consumer_pacts.yaml]: Removed `PACT_BROKER_URL`. Thanks for pointing this out. Didn't realize the url was already defined in `application=pacttest.properties`.

File: .github/actions/bump-skip/action.yml
Comment (ichengchang) [reply .github/actions/bump-skip/action.yml:4]: Code was created by @ddietterich from Verily so keeping the original authorship here.

File: .github/workflows/verify_consumer_pacts.yaml
Comment (ichengchang) [reply .github/workflows/verify_consumer_pacts.yaml]: Good catch! This has been corrected.

File: src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java
Comment (jladieu) [inline src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java]: This comment should probably be merged with the one above it, or outright overwrite it.

File: .github/workflows/verify_consumer_pacts.yaml
Comment (jladieu) [inline .github/workflows/verify_consumer_pacts.yaml]: Would it make sense to extract the `develop` here to use `RELEASE_BRANCHES`, defined earlier?  This might help call out an appropriate seam for setting the branch on a per-repo basis later on when this script needs to work for different repos.  I've seen repos that use both `main` and `develop` in DSP so I know this will need to change depending on context.

File: .github/workflows/verify_consumer_pacts.yaml
Comment (jladieu) [inline .github/workflows/verify_consumer_pacts.yaml]: Pacticipant might also be a good thing to extract as a variable, but this can wait until a refactoring pass that intends to DRY all this stuff up into reusable action.

**Review by jladieu — APPROVED** [review]
LGTM

File: .github/workflows/tag.yml
Comment (jladieu) [reply .github/workflows/tag.yml:69]: Agree this would be a good question to have clarity on!

Comment (ichengchang) [reply .github/workflows/tag.yml:69]: Yes, your statement makes sense. Here's a summary of what's happening.

1.     Multiple Developers and Commits in a PR:
- Multiple developers can collaborate on the same pull request (PR).
- The PR can have multiple commits.

2.     Sherlock and Helm Charts for `datarepo`:
- Sherlock doesn't synchronize Helm charts for the `datarepo` on PR commits
- The synchronization is only called upon during merge commit to `develop` branch

3.     Continuous Publishing and Verification of Contracts:
- You still want contracts from the PR to be continually published or verified.
- The tagging mechanism is introduced to create a ledger of published and verified contracts in Pact Broker. For PRs, the tag is appended with the commit hash. When the live tag is pushed to GitHub during merge commit, the contracts will be published or verified using the live tag without hash.
- This ledger allows developers to trace issues on both the consumer or provider side.

**Review by okotsopoulos — APPROVED** [review]
The description is really helpful -- thank you!

File: src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java
Comment (ichengchang) [reply src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java]: Updated the comment.

File: src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java
Comment (ichengchang) [reply src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java]: Consolidated and revised the comment. @jladieu feel free to copy the latest comment to your TODO on `workspacemanager` repo.

File: .github/workflows/verify_consumer_pacts.yaml
Comment (ichengchang) [reply .github/workflows/verify_consumer_pacts.yaml]: I just added a RELEASE_BRANCHES envvar.

File: .github/workflows/verify_consumer_pacts.yaml
Comment (ichengchang) [reply .github/workflows/verify_consumer_pacts.yaml]: Added PACTICIPANT envvar

**Review by okotsopoulos — CHANGES_REQUESTED** [review]
Hi Ivan, pulling back my approval because I see there's been some churn -- once the PR stabilizes, I want to make sure I understand those subsequent changes before merging if that's okay.

**Comment by ichengchang** [comment]
> Hi Ivan, pulling back my approval because I see there's been some churn -- once the PR stabilizes, I want to make sure I understand those subsequent changes before merging if that's okay.

Sounds good. We are looking at a `dependabot` issue in BPM so it's OK to hold off this until we know what's going on there.

**Comment by ichengchang** [comment]
> > Hi Ivan, pulling back my approval because I see there's been some churn -- once the PR stabilizes, I want to make sure I understand those subsequent changes before merging if that's okay.
> 
> Sounds good. We are looking at a `dependabot` issue in BPM so it's OK to hold off this until we know what's going on there.

@okotsopoulos We identified a `bumper` issue that affects repositories using `dependabot.yml` to update dependencies. Since `jade-data-repo` does not use `dependabot.yml` the current PR should be good. 

Here's the list of changes that were made

- `dry-run` was set to `true` for all events in `verify_consumer_pacts.yml` 
- added `release-branches` variable to `tag.yml`

**Review by okotsopoulos — APPROVED** [review]

**Pull request merged**
